oobleck/losses.py

```diff
@@ -192,4 +188,31 @@ def vicreg_cov_loss(z):
     "the regularization term that is the sum of the off-diagaonal terms of the covariance matrix"
     num_features = z.shape[1]*z.shape[2]  # TODO: move this out for speed.
     cov_z = torch.cov(rearrange(z, 'b c t -> ( c t ) b'))   
-    return off_diagonal(cov_z).pow_(2).sum().div(num_features)
\ No newline at end of file
+    return off_diagonal(cov_z).pow_(2).sum().div(num_features)
+
+
+class VICRegLoss(nn.Module):
+    """VICReg, https://arxiv.org/abs/2105.04906
+    Typically there will already be some loss in use to enforce some similarity and draw points together. VICReg offers a sort of "repulsion" regularization, across the batch dimension, 
+    to keep the latent space from collapsing.  
+    Intended to be connected to the latents in the middle of the autoencoder.. """
+
+    def __init__(self, gamma: float = 1., eps: float = 1.0e-4, weight: float = 1.) -> None:
+        super().__init__()
+        self.gamma, self.eps = gamma, eps
+        self.weight = weight
+
+    def forward(self, inputs: TensorDict) -> TensorDict:
+        z = inputs['latent']
+        # writing with self.... "hooks" to enable extraction of  individual losses for logging
+        self.var_loss = vicreg_var_loss(inputs, gamma=self.gamma, eps=self.eps)
+        self.cov_loss = vicreg_cov_loss(inputs)
+        loss = self.var_loss + self.cov_loss
+        inputs = accumulate_value(
+            inputs,
+            {
+                "vicreg_loss": self.weight * loss
```

Comment:
Should be "generator_loss". This line adds the vic_reg loss to the overall accumulated loss while the `inputs.update()` line below is used to return the individual named losses.